drivers/spi/spi_silabs_eusart.c

```diff
@@ -199,82 +617,194 @@ static int spi_silabs_eusart_shift_frames(EUSART_TypeDef *eusart,
-	} while (!ret && spi_silabs_eusart_transfer_ongoing(data));
+	if (spi_silabs_eusart_transfer_ongoing(data)) {
+		do {
+			ret = spi_silabs_eusart_shift_frames(eusart_config->base, data);
+		} while (!ret && spi_silabs_eusart_transfer_ongoing(data));
+	}
 
 	spi_context_cs_control(ctx, false);
 	spi_context_complete(ctx, dev, 0);
+
+	spi_silabs_eusart_pm_policy_put(dev);
+	return ret;
 }
 
 /* API Functions */
 static int spi_silabs_eusart_init(const struct device *dev)
 {
-	int err;
 	const struct spi_silabs_eusart_config *eusart_config = dev->config;
 	struct spi_silabs_eusart_data *data = dev->data;
+	int err;
 
-	err = pinctrl_apply_state(eusart_config->pcfg, PINCTRL_STATE_DEFAULT);
+	err = clock_control_on(eusart_config->clock_dev, 
+			       (clock_control_subsys_t)&eusart_config->clock_cfg);
 	if (err < 0) {
 		return err;
 	}
 
+	err = pinctrl_apply_state(eusart_config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0 && (err != -ENOENT)) {
+		/** The return number ENOENT means that no sleep state is defined
+		 * by the user in the DT.
+		 */
+		return err;
+	}
+
 	err = spi_context_cs_configure_all(&data->ctx);
 	if (err < 0) {
 		return err;
 	}
+
 	spi_context_unlock_unconditionally(&data->ctx);
 
-	return 0;
+	pm_device_init_off(dev);
+
+	err = pm_device_runtime_enable(dev);
+
+	return err;
 }
 
 static int spi_silabs_eusart_transceive(const struct device *dev, const struct spi_config *config,
-				       const struct spi_buf_set *tx_bufs,
-				       const struct spi_buf_set *rx_bufs)
+					const struct spi_buf_set *tx_bufs,
+					const struct spi_buf_set *rx_bufs)
 {
 	struct spi_silabs_eusart_data *data = dev->data;
-	uint16_t control = 0;
 	int ret;
 
 	spi_context_lock(&data->ctx, false, NULL, NULL, config);
 
-	ret = spi_silabs_eusart_configure(dev, config, &control);
-	if (ret < 0) {
+	ret = spi_silabs_eusart_configure(dev, config);
+
+	if (ret) {
 		spi_context_release(&data->ctx, ret);
 		return ret;
 	}
 
 	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
-	spi_silabs_eusart_xfer(dev, config);
 
+#ifdef CONFIG_SPI_SILABS_EUSART_DMA
+	if (is_dma_enabled_instance(dev)) {
+		ret = spi_silabs_eusart_xfer_dma(dev, config);
+	} else {
+#endif
+		spi_silabs_eusart_xfer_polling(dev, config);
+#ifdef CONFIG_SPI_SILABS_EUSART_DMA
+	}
+#endif
+	ret = spi_context_wait_for_completion(&data->ctx);
 	spi_context_release(&data->ctx, ret);
 
-	return 0;
+	return ret;
 }
 
 #ifdef CONFIG_SPI_ASYNC
 static int spi_silabs_eusart_transceive_async(const struct device *dev,
-					     const struct spi_config *config,
-					     const struct spi_buf_set *tx_bufs,
-					     const struct spi_buf_set *rx_bufs,
-					     struct k_poll_signal *async)
+					      const struct spi_config *config,
+					      const struct spi_buf_set *tx_bufs,
+					      const struct spi_buf_set *rx_bufs,
+					      spi_callback_t cb,
+					      void *userdata)
 {
-	return -ENOTSUP;
+	struct spi_silabs_eusart_data *data = dev->data;
```

Comment:
There is still unnecessary duplication between the polling and async implementations. You should be able to implement this as:
```c
int transceive(dev, config, tx, rx, async, cb, userdata)
{
    /* The actual work goes here in a single implementation */
}

int transceive_polling(dev, config, tx, rx)
{
    return transceive(dev, config, tx, rx, false, NULL, NULL);
}

#ifdef CONFIG_SPI_ASYNC
int transceive_async(dev, config, tx, rx, cb, userdata)
{
    return transceive(dev, config, tx, rx, true, cb, userdata);
}
#endif
```